go/tools/filter_exec/filter_exec.go

```diff
@@ -0,0 +1,111 @@
+/* Copyright 2016 The Bazel Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"go/build"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+)
+
+var (
+	cgo  = flag.Bool("cgo", false, "Sets whether cgo-using files are allowed to pass the filter.")
+	tags = flag.String("tags", "", "Only pass through files that match these tags.")
+
+	absMarker    = "-abs-"
+	filterMarker = "-filter-"
+	escapeMarker = "-escape-"
+)
+
+// runCommand goes through it's arguments filtering out source code files that do not match
+// the supplied build context, and expanding the current working directory where needed.
+// It then invokes the executable with the remaining result.
+func runCommand(bctx build.Context, executable string, input []string) error {
+	var err error
+	args := []string{}
+	unfiltered := 0
+	filtered := 0
+	for _, in := range input {
+		if strings.HasPrefix(in, escapeMarker) {
```

Comment:
I don't think this escaping is quite correct.

If only one verb is allowed, then the conditions checking `absMarker` and `filterMarker` below should be exclusive: only one should be matched, and we shouldn't examine any text after the match. I think it would be cleaner to have all three conditions together, combined with the conditions below. Something like this:

```go
if strings.HasPrefix(in, escapeMarker) {
  in = in[len(escapeMarker):]
} else if strings.HasPrefix(in, absMarker) {
  in = in[len(absMarker):]
  in, err := filepath.Abs(in)
  ...
} else if strings.HasPrefix(in, filterMarker) {
  in = in[len(filterMarker):]
  dir, base := filepath.Split(in)
  ...
}
args = append(args, in)
```

If more than one verb can be matched, we'd need a loop. `escapeMarker` should terminate that loop when matched.